### ISSUE TITLE
Investigate why SMG is unable to receive Salesforce Platform Events

### ISSUE THREAD

**Issue by MykytaDuniashev** [elasticio/salesforce-component-v2#70]
# Investigation

## Description

SMG reported that some of their flows are not receiving SF platform events. Other flows are successful at capturing PEs.
The main problem is that two of their flows stop receiving Salesforce platform events. There could be a few reasons:

1. There were no new events from Salesforce, which is not the case because after restarting the flow it will work.
2. There are some problems with their Salesforce instance, we are unable to investigate it since we don't have access.
3. Lost connection to the Salesforce. We deployed a [new version](https://github.com/elasticio/salesforce-component-v2/releases/tag/2.7.1) which contains a fix but looks like it is not working.

**RE-SF-User** flow regularly produces `Client is offline` errors. Logs contain multiple `Client is online` messages instead of a single one.

```
[2023-09-26 16:19:04.423]: Subscription is still running, waiting for new messages
[2023-09-26 16:21:42.791]: Client is offline
[2023-09-26 16:21:42.791]: Lets call processTrigger one more time
[2023-09-26 16:21:42.791]: Starting Subscribe to platform events Trigger
[2023-09-26 16:21:42.802]: Client is offline
[2023-09-26 16:21:42.802]: Lets call processTrigger one more time
[2023-09-26 16:21:42.802]: Starting Subscribe to platform events Trigger
[2023-09-26 16:21:42.949]: Client is online
[2023-09-26 16:21:42.983]: Client is online
[2023-09-26 16:21:43.021]: Streaming client created and ready
[2023-09-26 16:21:43.021]: Subscribed to PushTopic successfully
[2023-09-26 16:21:43.053]: Streaming client created and ready
[2023-09-26 16:21:43.053]: Subscribed to PushTopic successfully
[2023-09-26 16:21:47.863]: Client is online
[2023-09-26 16:21:47.880]: Client is online
```

- [RE-SF-User](https://app-smg.elastic.io/c/648056e5f8e19300120b43bb/w/6480b707f8e19300121f433e/designer/6491a7b326a06000129029e6/v/latest)
PEs of this flow are generated less frequently than other flows, in low numbers. F.e. if there are new hires in a week, there can be 10 platform events for those 10 new users. For another week, there can be no events. 
- [RE-SF-Campaign_Member_RE](https://app-smg.elastic.io/c/648056e5f8e19300120b43bb/w/6480b707f8e19300121f433e/designer/6491a6ec26a06000128ff18e/v/latest)
PEs of this flow are generated less frequently than other flows, in high numbers. F.e. once a week, 3000 PEs in the same minute.

Previous improvements: https://github.com/elasticio/salesforce-component-v2/issues/68.
- [Link](https://drive.google.com/file/d/1bYG3Yhb9fSWEWOBF_ZVAolq8vQe1a85k/view?usp=sharing) to a .txt file with SMG Salesforce limits.
- [Link](https://app.intercom.com/a/inbox/qgp8r7fl/inbox/shared/all/conversation/10951100098837) to the Intercom conversation.

## Component Version

2.7.1

## Workaround(s)

Use the [Get Updated Objects Polling ](https://docs.elastic.io/components/salesforce/triggers.html#get-updated-objects-polling) trigger instead of [Subscribe to platform events](https://docs.elastic.io/components/salesforce/triggers.html#subscribe-to-platform-events-trigger). Since they should also get when a **Campaign Member Object** is deleted - meaning that a **Contact** is removed from the **Campaign**. [Get Updated Objects Polling ](https://docs.elastic.io/components/salesforce/triggers.html#get-updated-objects-polling) trigger can't be used in [RE-SF-Campaign_Member_RE](https://app-smg.elastic.io/c/648056e5f8e19300120b43bb/w/6480b707f8e19300121f433e/designer/6491a6ec26a06000128ff18e/v/latest) flow.

**Comment by virrax** [comment]
checked basic network activity for first steps for the following flows: 
- questionable `6491a7b326a06000129029e6` and `6491a6ec26a06000128ff18e`
- normal one `64c7d4e8f85e7600124ebb31`


for the normal one:
```
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 10.31.0.125:39584       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.0.125:53238       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.0.125:42206       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.0.125:35184       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.0.125:35870       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.0.125:39146       ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.0.125:39144       ec2-18-170-131-54:https ESTABLISHED
```

for the first `questionable`:
```
tcp        0      0 10.31.47.60:47228       ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.47.60:60232       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.47.60:35878       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:57272       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.47.60:53040       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:42054       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.47.60:38528       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:55294       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:48928       ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.47.60:39858       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:60442       ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.47.60:46408       ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.47.60:36660       ec2-13-41-16-236.:https ESTABLISHED
```

for the second `questionable`:
```
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 10.31.5.25:58264        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:37036        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:57590        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:50510        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:59068        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:56960        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:50604        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:45548        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:51314        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:47500        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:51286        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:50020        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:53948        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:57534        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:50018        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:39198        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:49624        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:53942        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:55130        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:50896        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:45810        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:50124        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:40298        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:45534        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:59602        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:47512        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:35110        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:36450        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:45544        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:57474        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:43188        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:53558        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:39764        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:35236        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:59082        ec2-18-170-131-54:https ESTABLISHED
tcp        0      0 10.31.5.25:54450        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:45822        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:57528        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:36440        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:43172        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:59418        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:43204        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:49622        ec2-18-170-190-12:https ESTABLISHED
tcp        0      0 10.31.5.25:51576        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:40806        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:40280        ec2-13-41-16-236.:https ESTABLISHED
tcp        0      0 10.31.5.25:36442        ec2-18-170-190-12:https ESTABLISHED
```

the only pattern I see - working one definitely has less established connections

maybe, just maybe we have an issue with rate-limiting or stuck connections

**Comment by MykytaDuniashev** [comment]
Krombacher noticed the duplicate PE messages after the first error appeared in their flows with v2.7.1. They will wait and update their flows to the latest version when we fix this. I'll keep them posted. 
[Link](https://app.intercom.com/a/inbox/qgp8r7fl/inbox/shared/all/conversation/10951100098981) to the Intercom conversation.

**Comment by MykytaDuniashev** [comment]
Many thanks to @if0s, @HannaTrotsenko ! We have positive [feedback](https://app.intercom.com/a/inbox/qgp8r7fl/inbox/shared/all/conversation/10951100098837?view=List) from SMG:
```
I deployed the new component version on all flows on October 9th evening.
On October 10th, 11:36, 1170 add messages integrated successfully.
On October 10th, 11:26, 4585 add messages integrated successfully.
On October 16th, 09:04, 4278 add messages integrated successfully.
We can see that the flow was sleeping for almost 6 days but it didn't die out. 
So I deduce that the bugfix is successful. 
I shared these findings internally and with the customer as well. 
So once again, thanks for your tight follow-up and thanks to the product team! 
I think we achieved an important fix for every party!
```
Let's wait for feedback from Krombacher to be sure that it works for them too.

**Comment by MykytaDuniashev** [comment]
Resolved

**Comment by HannaTrotsenko** [comment]
@MykytaDuniashev pls don't close such issues - I didn't add it into changelod btcause it closed.

**Comment by HannaTrotsenko** [comment]
https://github.com/elasticio/salesforce-component-v2/releases/tag/2.7.2

**Comment by MykytaDuniashev** [comment]
The ticket reopened since the confirmation that the issue has been fixed came too early, the colleague confirmed it is no longer working.

> We still experience issues with the flows that are dependent on the Salesforce Platform Events. My former colleague Ardan is no longer working for us (valantic), therefore I can not access his support ticket history.
> 
> **Problem description:**
> Basically, the issue is the same as last time. We are using the Salesforce Platform Event connector to subscribe to Platform events coming from Salesforce. If such an Elastic flow has not been triggered for a longer time (approximately a few days), the flow “dies”. So, when a new Platform Event is coming from Salesforce, the Elastic flow is not triggered anymore. This has led to data inconsistencies between Salesforce and Bloomreach.
>
> You can find an example of such a flow here: https://app-smg.elastic.io/c/648056e5f8e19300120b43bb/w/648059af2ae0600012320ec9/designer/648068f9be39e400129c4e67/v/latest

**Comment by MykytaDuniashev** [comment]
@if0s Looks like our fix is not working
![image](https://github.com/elasticio/salesforce-component-v2/assets/108279772/224aa063-c55f-440f-8ec0-12e6b7efe014)

**Comment by andrewreshitko** [comment]
Fixed in #76
